### PR TITLE
Try out ProxyManager

### PR DESCRIPTION
**Pull request by wouterj**
I started trying out some of the ideas in #110 and #113.

The idea of this change is that instead of a "fake" `Proxy` object, we return a proxied entity/model. This allows removing `->object()` calls and removing some confusion when trying to pass the objects into typehinted functions. For this to work, the solution has to fulfill these promises:

* [x] The proxied entity/model has to support auto-refresh at least as good as the current `Proxy`
* [x] The AR helpers (`save()`, `remove()`, ...) have to be moved to `Factory`/`ModelFactory`
* [x] In 1.x, this new behavior should be opt-in per factory. This allows a smooth migration path, before removing the old factories in 2.0
* [ ] Reintroduce support for global auto refresh setting
* [ ] Document UPGRADE guide + deprecate the old Proxy

### CODE REVIEW

File: tests/Unit/ProxyGeneratorTest.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        $proxyTag->setName('proxy');
+        $this->assertEquals('proxy', $proxyTag->getName());
+    }
+
+    /** @test */
+    public function can_only_auto_refresh_getters(): void
+    {
+        $tag = new Tag();
+        $this->managerRegistry->expects($this->any())->method('getManagerForClass')->with(Tag::class)->willReturn($this->objectManager);
+        $proxyTag = $this->generator->generate($tag, ['get*']);
```
Comment (wouterj) [inline tests/Unit/ProxyGeneratorTest.php:51]: This is my try at "working around" the auto-refresh problems from #113. You can specify which methods should be proxied/use auto-refresh (allowing for `*` globs, like `get*` or `['get*', 'has*, 'is*']`). The idea is that you will be able to configure this in the model factory.

I'm not sure if this is the ideal solution, but I didn't really understand the `spl_object_id()` solution proposed in #113 😇

File: src/Instantiator.php [hunk 2]
```diff
@@ -123,6 +126,16 @@ public function alwaysForceProperties(array $properties = []): self
         return $this;
     }
 
+    /**
+     * Instead of returning an instance of {@see Proxy}, use a generator to create a real proxy for the object.
+     */
+    public function withProxyGenerator(): self
```
Comment (wouterj) [inline src/Instantiator.php:132]: My original idea was using this to enable the proxy generator. However, I'm not sure if this should live in `Instantiator` or `Factory`. It's much more common in Foundry to modify `Instantiator` rather than the factory instance. On the other hand, `Factory` is currently creating the `Proxy` and the `ProxyGenerator` needs access to the factory's `Configuration` instance.

File: tests/Unit/ProxyGeneratorTest.php (hunk 1 above)
Comment (kbond) [reply tests/Unit/ProxyGeneratorTest.php:51]: I can't tell if this solves the issue in #113. Might be better to test on an entity with multiple fields like `Post`.

> but I didn't really understand the spl_object_id()

Long story short, I don't think this will work. Basically, I was trying to come up with a way to only refresh the object if it's been changed in the db.

File: src/Instantiator.php (hunk 2 above)
Comment (kbond) [reply src/Instantiator.php:132]: Hmm, my first thought is `Factory` as it requires the `Configuration` as you say. Obviously given the static nature of everything it doesn't really matter but I've been trying to follow proper a dependency injection pattern where possible.

File: tests/Unit/ProxyGeneratorTest.php (hunk 1 above)
Comment (wouterj) [reply tests/Unit/ProxyGeneratorTest.php:51]: Basically, my idea was that a user can limit which methods will cause a refresh. E.g. you wouldn't necessarily refresh on setters, which automatically resolves the issue when calling multiple setters.

We can maybe also make this more "magic" by running the fresh post-method invocation and only run it for methods that return something (and then, after a refresh, call the method again on the entity from within the interceptor). However, not sure if there are edge-cases there (e.g. fluent setters).

Comment (kbond) [reply tests/Unit/ProxyGeneratorTest.php:51]: I had thought about limiting to non-setters (and of course make configurable) but for models that have rich domains, this becomes a burden I think. I came to the conclusion that it's better to refresh everything and give the ability to disable auto-refreshing for certain circumstances which I believe are rare.

I'm relying on auto-refreshing (on by default) in 100's of tests and only have one instance where I have to disable it.

File: tests/Functional/ProxyGeneratorTest.php [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        $this->markTestSkipped('not supported');
+    }
+
+    /**
+     * @test
+     */
+    public function can_force_set_and_save(): void
+    {
+        $this->markTestSkipped('not supported');
+    }
```
Comment (wouterj) [inline tests/Functional/ProxyGeneratorTest.php:55]: I did not move the `forceSet`/`forceGet` methods to the factory (yet). Is this a feature we really want to keep in Foundry?

File: tests/Functional/ProxyGeneratorTest.php [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+{
+    protected static $POST_FACTORY = PostFactoryWithProxyGenerator::class;
+
+    /**
+     * @test
+     */
+    public function can_assert_persisted(): void
+    {
+        $this->markTestSkipped('not supported');
+    }
```
Comment (wouterj) [inline tests/Functional/ProxyGeneratorTest.php:23]: TODO: we probably have to move the `assertPersisted()` to the assertions class

File: tests/Functional/ProxyGeneratorTest.php [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        $this->assertSame('old title', $post->getTitle());
+        $this->assertSame('old body', $post->getBody());
+
+        // reboot kernel
+        static::ensureKernelShutdown();
+        static::bootKernel();
+
+        $this->assertSame('old title', $post->getTitle());
+        $this->assertSame('old body', $post->getBody());
+    }
```
Comment (wouterj) [inline tests/Functional/ProxyGeneratorTest.php:100]: TODO: I'm not sure if it's possible to replace the value holder instance. This is required for re-fetching of entities (e.g. after the entity manager is cleared).

**Review by wouterj — COMMENTED** [review]
After a very long silence, I've made some progress with this feature today.

Auto refresh is now implemented exactly like in the current `Proxy` object and other helper functions are moved to the factory. The implementation is not yet feature-complete compared to the old `Proxy`. I've left some comments on the missing bits (some of which I'm not sure if it's possible to implement).

**Comment by wouterj** [comment]
Btw, I can't reproduce the Psalm errors locally :confused:

File: tests/Functional/ProxyGeneratorTest.php (hunk 3 above)
Comment (kbond) [reply tests/Functional/ProxyGeneratorTest.php:55]: I do find them useful for feature-rich models with no public access to certain fields (ie verification or password reset token).

I'd be fine with leaving them on the `Instantiator` or adding some helper functions.

File: tests/Functional/ProxyGeneratorTest.php (hunk 4 above)
Comment (kbond) [reply tests/Functional/ProxyGeneratorTest.php:23]: I don't personally find the Proxy assertions useful, maybe we can remove in 2.0?

**Review by kbond — COMMENTED** [review]
Awesome, thanks for working on this @wouterj! Couple questions/thoughts:
1. Am I correct in that we can't add methods to the generated proxy?
2. Are you planning a way to enable globally via config?

**Comment by wouterj** [comment]
> Am I correct in that we can't add methods to the generated proxy?

Yes. Proxies, at least within the scope of the ProxyManager library, are meant to behave exactly like the real object (with the proxy thing being an implementation detail).
The proxy manager has a couple proxy types: value holder (used here, allows you to call code before and after each real object method call); and lazy loading proxies (those you know from Doctrine, where the entity is only fetched from the db when you call a method).

> Are you planning a way to enable globally via config?

I think it should be doable. I made the opt-in per factory at the moment, as that allows you to migrate one factory at the time. Also, it means that e.g. third party factories can already add support for it while your application's factories don't.

File: tests/Functional/ProxyGeneratorTest.php (hunk 4 above)
Comment (wouterj) [reply tests/Functional/ProxyGeneratorTest.php:23]: That sounds great :)

Honestly, I personally would like to remove as much custom methods from the proxy as possible. I much rather use the normal Doctrine API. But I know you're quite fond of these little helper methods. In short: you're completely at the lead as to which proxy methods to keep and which to drop in 2.0 :)

**Comment by wouterj** [comment]
To summarize our Slack conversation:

* [ ] Move these proxy methods to normal functions: `force_set`, `force_get`, `disable_autorefresh`, `enable_autorefresh`, `without_autorefresh`?
* [ ] Move `save` and `remove` to methods on the `Factory`
* [ ] Handling auto refresh while the kernel was rebooted is a must have

**Comment by wouterj** [comment]
I've updated the PR solving the biggest blocker: refreshing the model (value holder) of the proxy.

I'll try to find some time later this week/weekend to fix the 2 remaining (simple) tasks. After that, this should be ready :)

File: src/Proxy/ProxyGenerator.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+                $objectManager->refresh($model);
+
+                return;
+            }
+
+            // refetch the model as it's no longer managed
+            $modelId = $modelClassMetadata->getIdentifierValues($model);
+            if (!$modelId) {
+                // TODO: throw exception?
```
Comment (wouterj) [inline src/Proxy/ProxyGenerator.php:56]: The current `Proxy` class sets the value holder to `null`. Is that what we want? Or should we throw an exception? Or leave the value holder as-is?

Comment (kbond) [reply src/Proxy/ProxyGenerator.php:55]: If I understand correctly, currently, if the object cannot be "refreshed", we [throw an exception](https://github.com/zenstruck/foundry/blob/d2cd4c7b4c6821b96d8fda0c1d2febad345ffa35/src/Proxy.php#L167).

**Comment by kbond** [comment]
> Btw, I can't reproduce the Psalm errors locally

https://github.com/laminas/laminas-code/issues/67 looks similar.

File: src/Proxy/ProxyGenerator.php (hunk 6 above)
Comment (wouterj) [reply src/Proxy/ProxyGenerator.php:55]: Oh, I misread the code. Exception it is :)

> By "value holder", do you mean this? 

Yes, in ProxyManager terms, "value holder" is the object that is being proxied.

File: README.md [hunk 7]
```diff
@@ -20,7 +20,7 @@ $post = PostFactory::new() // Create the factory for Post objects
 The factories can be used inside [DoctrineFixturesBundle](https://symfony.com/doc/master/bundles/DoctrineFixturesBundle/index.html)
 to load fixtures or inside your tests, [where it has even more features](#using-in-your-tests).
 
-Want to watch a screencast 🎥 about it? Check out https://symfonycasts.com/foundry
+Want to watch a screencast  about it? Check out https://symfonycasts.com/foundry
```
Comment (wouterj) [inline README.md:23]: This should probably be reverted while merging (my terminal has a hard crash when it tries to display emojis currently, making it impossible to edit this file)

**Comment by wouterj** [comment]
PR updated once again with the helper functions and an updated README. As the proxy object now really behaves like the real object, I've removed most of the mentions of Proxy (it's mostly an implementation detail now).

While documenting, I discovered 2 remaining tasks that I've added to the PR description.

File: README.md (hunk 7 above)
Comment (kbond) [reply README.md:23]: I'll leave it out. If it causes an issue for you it will for others as well.

**Comment by kbond** [comment]
Awesome, thanks for continuing to work on this!  Have you thought anymore a out how we're going to migrate users to this? It'd be nice if new installs auto used this and existing installs have to choose one way or another (like what we did in #131). The maker should be updated as well.

**Comment by wouterj** [comment]
> Have you thought anymore a out how we're going to migrate users to this?

Currently, you would opt-in for the new proxy manager in each factory:

```php
class PostFactory extends ModelFactory
{
    protected function initialize()
    {
        return $this->withProxyGenerator();

        // or, if you want to limit proxy to only some methods (supporting globs)
        //return $this->withProxyGenerator(['get*']);
    }
}
```

I suggest adding a deprecation notice in `Factory#create()` if auto refresh is `true` and the proxy generator was not enabled (i.e. the old proxies are in use). The `make:factory` maker should indeed be updated to use `->withProxyGenerator()` by default.

Optionally, we can add a global option as you suggested. The advantage would be an easier migration when we remove the `withProxyGenerator()` method (probably in Foundry 3, if you stick to Symfony's BC policy). The disadvantage is that this disallows updating one factory each time (when updating the docs, I realized it's quite some things you have to change). I'm happy to hear if you would like such a global option :)

**Comment by kbond** [comment]
I'm on board 100% with regards to `ModelFactory`'s and giving the ability to upgrade on a factory-by-factory basis. With `AnonymousFactory`'s it will be tricky without a global option. We certainly don't want users to have to call `withProxyGenerator()` for these.

Following Symfony's BC policy, we can't remove the current Proxy system entirely in 2.0?

**Comment by kbond** [comment]
What about hooking into the current `auto_refresh_proxies` config and making it an enum: `new`, `old`, `disable` (need help with better names :thinking:) and make using a boolean deprecated?

**Comment by wouterj** [comment]
> Following Symfony's BC policy, we can't remove the current Proxy system entirely in 2.0?

Introducing a new system is a 3 step progress in Symfony BC:

1. In a *minor version* (`x.y`): Introduce the new system, add an option/method to opt in to the new system, and deprecate not doing the opt-in (i.e. trigger deprecation if `withProxyManager()` isn't called)
2. On the next *major version* (`(x+1).0`): switch the opt-in to opt-out and throw an exception if the new system isn't opt-in (i.e. `withProxyManager()` isn't called). You can safely remove the old system now.
3. On the next *major version* (`(n+2).0`): remove the opt-in.

If you remove the method directly in step (2), you require people to update their code to remove the opt-in. This is also why I think it's very strict and OK for smaller libraries to not do this (the method is a useless bit of code hanging around for the entirety of Foundry 2).


> What about hooking into the current auto_refresh_proxies config and making it an enum: new, old, disable (need help with better names thinking) and make using a boolean deprecated?

Hmm, this is an interesting idea. This would allow you to change the "default" and change it on a factory basis using `withProxyManager(bool $enabled = true)`. This is the best of 2 worlds: an app that is fully migrated wouldn't have `withProxyManager()` anywhere and everyone is free to update each factory individually.

I think adding a new boolean option would make the most sense ("new" doesn't mean much in Foundry 2). What about `auto_refresh: true/false`?

**Comment by kbond** [comment]
> If you remove the method directly in step (2), you require people to update their code to remove the opt-in. This is also why I think it's very strict and OK for smaller libraries to not do this (the method is a useless bit of code hanging around for the entirety of Foundry 2).

Yeah, this does seem a bit strict. I'd like to remove `Proxy` (and related code) completely in 2.0. Do you agree? If so, `withProxyManager` wouldn't exist in 2.0, correct? I'd hate to add this to the maker, then deprecated and remove in 2.0. Or maybe I'm not fully understanding?

> I think adding a new boolean option would make the most sense ("new" doesn't mean much in Foundry 2). What about auto_refresh: true/false?

So what would happen if `auto_refresh: true` and `auto_refresh_proxies: true`? Error?

**Comment by kbond** [comment]
Something that just occurred to me: public properties cannot be auto-refreshed. I don't know how common it is to use public properties on entities. I imagine with static analysis becoming more popular and read-only properties in 8.1 it will become more common.

**Comment by kbond** [comment]
I've been playing with an alternate implementation using Symfony 6.2's `ProxyHelper` (and a bit of hacking). Here's my POC: https://github.com/kbond/symfony-reproducer/tree/foundry-real-proxy 

Here's the relevant [commit](https://github.com/kbond/symfony-reproducer/commit/abf7842c61c8409b0728914efb7b2dab3f45030a) and files:
- https://github.com/kbond/symfony-reproducer/blob/foundry-real-proxy/src/ProxyFactory.php (this is what foundry's Factory could look like)
- https://github.com/kbond/symfony-reproducer/blob/foundry-real-proxy/src/Proxy.php (the generated proxy implements this interface)
- https://github.com/kbond/symfony-reproducer/blob/foundry-real-proxy/src/IsProxy.php (the generated proxy uses this trait which implements the above interface)
- https://github.com/kbond/symfony-reproducer/blob/foundry-real-proxy/tests/RealProxyTest.php (the test that proves this works - everything is fully auto-completable in phpstorm)

Auto-refreshing doesn't work - maybe this is possible with some more tweaks. I'm also ok deprecating auto-refreshing and removing in 2.0. It is complex and has caused some confusion...

/cc @nikophil

**Comment by wouterj** [comment]
> Auto-refreshing doesn't work - maybe this is possible with some more tweaks. I'm also ok deprecating auto-refreshing and removing in 2.0. It is complex and has caused some confusion...

In my honest opinion, if you plan on deprecating auto-refreshing I would advise to completely deprecate the proxy stuff. It's too complex and maintenance heavy to introduce a concept (AR) to entities that Symfony developers aren't really used to use.

I would then propose to just apply the changes of this PR related to moving the AR helpers to `Factory`: https://github.com/zenstruck/foundry/pull/127/files#diff-f1158fbec0bb446d87dbd5346fba683c58ccec1a98b1787511ec7786585e8c31

**Comment by kbond** [comment]
> In my honest opinion, if you plan on deprecating auto-refreshing I would advise to completely deprecate the proxy stuff. It's too complex and maintenance heavy to introduce a concept (AR) to entities that Symfony developers aren't really used to use.

That's a fair point. Although, current users of foundry _are_ used to the AR concept (assuming they're using these methods on the current `Proxy` object).

I'm not sure yet about deprecating auto-refreshing... When it works, it's quite nice.

**Comment by kbond** [comment]
My [POC](https://github.com/kbond/symfony-reproducer/commit/d72cc234acfaef1357dab3994d83e080ab272ba4) now includes auto-refreshing methods (not public properties):

```php
$post->_enableAutoRefresh();

$post->getTitle(); // auto-refreshes before returning value
$post->body; // does not auto-refresh...
```
